### PR TITLE
New Adapter: Brave

### PR DESCRIPTION
**Pull request by thebraveio**

### CODE REVIEW

**Review by VeronikaSolovei9 — APPROVED** [review]
LGTM!

File: adapters/brave/brave.go
Comment (VeronikaSolovei9) [inline adapters/brave/brave.go:40]: Optimization suggestion: if there are more than one impressions in request this code will be executed for every imp, and `braveExt` will have data from the last imp in request. Do you think it makes sense to move it outside of the for loop and execute `getImpressionExt` just once?

File: adapters/brave/bravetest/supplemental/invalid-response.json
Comment (VeronikaSolovei9) [inline adapters/brave/bravetest/supplemental/invalid-response.json:1]: Super nitpick: please remove empty lines from json files where they are present.

**Comment by bretg** [comment]
docs PR https://github.com/prebid/prebid.github.io/pull/4039

**Comment by AlexBVolcy** [comment]
There is a PR https://github.com/prebid/prebid-server/pull/2370 that is being merged today that will cause conflicts to new adapter PRs. 

To resolve these conflicts, you'll need to do the following:

In `brave.go`, update the `Builder()` function signature to include a new parameter `server config.Server`

In `brave_test.go`, update any test that includes a call to `Builder()` to pass a `config.Server` object, to that call (i.e. `config.Server{ExternalUrl: "http://hosturl.com", GvlID: 1, DataCenter: "datacenter"}`)

After the merge, you can look at other adapters to see how the server object is implemented as an example.

File: adapters/brave/brave.go
Comment (SyntaxNode) [inline adapters/brave/brave.go:46]: With this code, the braveExt from the last imp is the only one which matters. It looks like your server includes the placement id in the url and thus only supports one placement per request? If so, please consider impression splitting: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#impression-splitting

File: adapters/brave/brave.go
Comment (SyntaxNode) [inline adapters/brave/brave.go:160]: Although many adapters do this, we consider it an anti-pattern. Does your bidding server return the media type in the response? Such as the new OpenRTB 2.6 `mtype` field?

File: adapters/brave/brave_test.go
Comment (SyntaxNode) [inline adapters/brave/brave_test.go:14]: Please consider using an obviously fake url for testing. There is no need for this to match the real endpoint and might be harder to maintain if you need to change it in the future.

File: exchange/adapter_builders.go
Comment (SyntaxNode) [inline exchange/adapter_builders.go:208]: Thank you for keeping the lists alphabetical. :)

**Review by SyntaxNode — COMMENTED** [review]
Nice test coverage. Left a few comments about multi impression support and bid type detection.

**Comment by SyntaxNode** [comment]
This code is identical to https://github.com/prebid/prebid-server/pull/2399. If these are aliases of the same underlying bidder, it's better to implement it just once and use an alias. Less code to maintain.

**Comment by bsardo** [comment]
> This code is identical to #2399. If these are aliases of the same underlying bidder, it's better to implement it just once and use an alias. Less code to maintain.

Hi @thebraveio, just checking in to see if you can shed some light on this. Thanks!

File: adapters/brave/brave.go
Comment (thebraveio) [reply adapters/brave/brave.go:46]: yes guys you right, removed it as not required (first one rewrite on loop not make sense, also we using only one impression array index on backends)

File: adapters/brave/brave.go
Comment (thebraveio) [reply adapters/brave/brave.go:160]: no, we using 2.5 and response with adm without mtype, that's why we have to recognise type by using req parameter

File: adapters/brave/brave_test.go
Comment (thebraveio) [reply adapters/brave/brave_test.go:14]: oh ok, got it , thanks

File: exchange/adapter_builders.go
Comment (thebraveio) [reply exchange/adapter_builders.go:208]: for sure, do not want to mess up your clean code

File: adapters/brave/brave.go
Comment (thebraveio) [reply adapters/brave/brave.go:40]: > > This code is identical to #2399. If these are aliases of the same underlying bidder, it's better to implement it just once and use an alias. Less code to maintain.
> 
> Hi @thebraveio, just checking in to see if you can shed some light on this. Thanks!

Yes , I developed code for both, but its different companies and strategy, not sure I will support both of them in future

Comment (thebraveio) [reply adapters/brave/brave.go:40]: > > This code is identical to #2399. If these are aliases of the same underlying bidder, it's better to implement it just once and use an alias. Less code to maintain.
> 
> Hi @thebraveio, just checking in to see if you can shed some light on this. Thanks!

Yes , I developed code for both, but its different companies and strategy, not sure I will support both of them in future

**Comment by thebraveio** [comment]
Added commit with fixes

**Comment by bsardo** [comment]
Hi @thebraveio, can you merge with master when you get a chance? Your code is failing the go 1.19 validation checks due to `go fmt` changes that were introduced with the recent upgrade to go 1.19 on master.

Also your code is failing the `validate-merge` step due to some recent changes we made on master regarding the ORTB library. You'll need to update the reference to the mxmcherry version of the library to the prebid version.

**Comment by thebraveio** [comment]
updated, check it again pls

**Comment by bsardo** [comment]
@thebraveio you still have errors to fix. The tests aren't passing because you have references to video heroes data structures in this PR.

**Comment by thebraveio** [comment]
> @thebraveio you still have errors to fix. The tests aren't passing because you have references to video heroes data structures in this PR.

yes, my mistake, sorry

**Comment by thebraveio** [comment]
hi Guys, could you review the pull ? it just on pending for review

File: go.sum
Comment (VeronikaSolovei9) [inline go.sum:32]: Please do not push changes in go.sum if you didn't modify go.mod. This should not be a part of this PR, there are no new libraries added

**Comment by VeronikaSolovei9** [comment]
Local testing looks good. Left minor comment about go.sum

**Review by VeronikaSolovei9 — DISMISSED** [review]

**Review by AlexBVolcy — APPROVED** [review]
LGTM

**Comment by bsardo** [comment]
Hi @thebraveio, would you mind rolling back the changes you made to go.sum?

**Comment by bsardo** [comment]
Hi @thebraveio, just checking in to see if you saw my earlier request to roll back changes to go.sum. No rush 🙂

**Comment by videoheroes** [comment]
sorry for a delay, I checkout file to version was pulled before push

**Comment by bsardo** [comment]
@videoheroes, @thebraveio, no worries. It looks like you have two more lines in `go.sum` that involve the `github.com/go-sql-driver/mysql` package to restore and that is why the tests are failing.

**Comment by videoheroes** [comment]
yes, seems like I need to update go.*  to https://github.com/prebid/prebid-server/pull/2356)

**Review by AlexBVolcy — APPROVED** [review]
LGTM! Discussed with the team today, you don't have to worry about the Trivy security failure, as we can still approve and merge in a PR if that's the only thing wrong with it.

**Review by VeronikaSolovei9 — APPROVED** [review]
LGTM

**Pull request merged**
